drivers/serial/uart_nrfx_uarte.c

```diff
@@ -2109,36 +2111,47 @@ static int uarte_nrfx_pm_action(const struct device *dev,
 		      &uart_nrfx_uarte_driver_api)
 
 #define UARTE_CONFIG(idx)						       \
+	.char_out = &uarte##idx##_char_out,				       \
+	.rx_data = &uarte##idx##_rx_data,				       \
 	.uart_config = {						       \
 		.baudrate = UARTE_PROP(idx, current_speed),		       \
 		.data_bits = UART_CFG_DATA_BITS_8,			       \
 		.stop_bits = UART_CFG_STOP_BITS_1,			       \
 		.parity = IS_ENABLED(CONFIG_UART_##idx##_NRF_PARITY_BIT)       \
 			  ? UART_CFG_PARITY_EVEN			       \
 			  : UART_CFG_PARITY_NONE,			       \
-		.flow_ctrl = UARTE_PROP(idx, hw_flow_control)  \
+		.flow_ctrl = UARTE_PROP(idx, hw_flow_control)		       \
 			     ? UART_CFG_FLOW_CTRL_RTS_CTS		       \
 			     : UART_CFG_FLOW_CTRL_NONE,			       \
 	}
 
-#define UARTE_ASYNC(idx)						       \
-	IF_ENABLED(CONFIG_UART_##idx##_ASYNC,				       \
-		(struct uarte_async_cb uarte##idx##_async = {		       \
-			.hw_rx_counting =				       \
-				IS_ENABLED(CONFIG_UART_##idx##_NRF_HW_ASYNC),  \
+#define UARTE_ASYNC(idx)							       \
+	IF_ENABLED(CONFIG_UART_##idx##_ASYNC,					       \
+		(static uint8_t uarte##idx##_tx_cache[CONFIG_UART_ASYNC_TX_CACHE_SIZE] \
+			UARTE_MEMORY_SECTION(idx);				       \
+		struct uarte_async_cb uarte##idx##_async = {			       \
+			.tx_cache = uarte##idx##_tx_cache,			       \
+			.hw_rx_counting =					       \
+				IS_ENABLED(CONFIG_UART_##idx##_NRF_HW_ASYNC),	       \
 		}))
 
 #define UARTE_INT_DRIVEN(idx)						       \
-	IF_ENABLED(CONFIG_UART_##idx##_INTERRUPT_DRIVEN,	       \
-		(static uint8_t uarte##idx##_tx_buffer[\
-			MIN(CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE,	       \
-			    BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE))];      \
+	IF_ENABLED(CONFIG_UART_##idx##_INTERRUPT_DRIVEN,		       \
+		(static uint8_t uarte##idx##_tx_buffer			       \
+			[MIN(CONFIG_UART_##idx##_NRF_TX_BUFFER_SIZE,	       \
+			    BIT_MASK(UARTE##idx##_EASYDMA_MAXCNT_SIZE))]       \
+			    UARTE_MEMORY_SECTION(idx);			       \
 		 static struct uarte_nrfx_int_driven			       \
 			uarte##idx##_int_driven = {			       \
 				.tx_buffer = uarte##idx##_tx_buffer,	       \
 				.tx_buff_size = sizeof(uarte##idx##_tx_buffer),\
 			};))
 
+#define UARTE_MEMORY_SECTION(idx)					       \
+	COND_CODE_1(UARTE_HAS_PROP(idx, memory_region),			       \
+	(__attribute__((__section__(LINKER_DT_NODE_REGION_NAME(		       \
+	UARTE_DMA_MEMORY_REGION(idx)))))), ())
```

Comment:
I'd suggest to get rid of `UARTE_DMA_MEMORY_REGION()` and move this definition into its place. Then it will be easier to see what it is all about at first glance, I think.
```suggestion
#define UARTE_MEMORY_SECTION(idx)					       \
	COND_CODE_1(UARTE_HAS_PROP(idx, memory_region),			       \
		(__attribute__((__section__(LINKER_DT_NODE_REGION_NAME(	       \
			DT_PHANDLE(UARTE(idx), memory_region)))))),	       \
		())
```